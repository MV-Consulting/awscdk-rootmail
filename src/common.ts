const integTestWildcard = '${Token[AWS.Region';
// see https://aws.amazon.com/de/about-aws/whats-new/2023/09/amazon-ses-email-service-7-regions/
export const sesEnabledRegions = [
  'us-east-1', // US East (N. Virginia),
  'eu-west-1', // Europe (Ireland),
  'us-west-2', // US West (Oregon),
  // new regions since 2023-09
  'eu-central-1', // Europe (Frankfurt),
  'eu-west-2', // Europe (London),
  'us-east-2', // US East (Ohio),
  'ca-central-1', // Canada (Central),
  'ap-northeast-1', // Asia Pacific (Tokyo),
  'ap-southeast-1', // Asia Pacific (Singapore),
  'ap-southeast-2', // Asia Pacific (Sydney),
];

export function isSESEnabledRegion(region: string): boolean {
  return sesEnabledRegions.includes(region) || region.startsWith(integTestWildcard);
}