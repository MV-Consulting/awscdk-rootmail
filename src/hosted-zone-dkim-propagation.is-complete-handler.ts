import { SES } from '@aws-sdk/client-ses';
// eslint-disable-next-line import/no-unresolved
import * as AWSCDKAsyncCustomResource from 'aws-cdk-lib/custom-resources/lib/provider-framework/types';
export const PROP_DOMAIN = 'Domain';

const ses = new SES();

export interface IsCompleteHandlerResponse {
  IsComplete: boolean;
}

export async function handler(event: AWSCDKAsyncCustomResource.OnEventRequest): Promise<IsCompleteHandlerResponse> {
  const domain = event.ResourceProperties[PROP_DOMAIN];

  switch (event.RequestType) {
    case 'Create':
      const isReady = await internalHandler(domain);
      return { IsComplete: isReady };
    case 'Update':
    case 'Delete':
      return {
        IsComplete: true,
      };
  }
}

function log(msg: any) {
  console.log(JSON.stringify(msg));
}

async function internalHandler(domain: string): Promise<boolean> {

  log({
    domain: domain,
    level: 'debug',
  });

  const sendingResponse = await ses.getAccountSendingEnabled({});
  if (!sendingResponse.Enabled) {
    return false;
  }
  log('sending enabled');

  const identityVerificationResponse = await ses.getIdentityVerificationAttributes({ Identities: [domain] });
  const identityVerificationStatus = identityVerificationResponse.VerificationAttributes![domain].VerificationStatus;
  if (identityVerificationStatus !== 'Success') {
    log(`Identity Verification status not successful. Was '${identityVerificationStatus}'`);
    return false;
  }
  log('identitity verification successful');

  const identityDkimRes = await ses.getIdentityDkimAttributes({ Identities: [domain] });
  const identityDkimStatus = identityDkimRes.DkimAttributes![domain].DkimVerificationStatus;
  if (identityDkimStatus !== 'Success') {
    log(`DKIM status not successful. Was '${identityDkimStatus}'`);
    return false;
  }
  log('DKIM verification successful');

  const identityNotificationRes = await ses.getIdentityNotificationAttributes({ Identities: [domain] });
  const forwardingEnabled = identityNotificationRes.NotificationAttributes![domain].ForwardingEnabled;
  if (!forwardingEnabled) {
    log(`Forwarding not enabled. Was '${forwardingEnabled}'`);
    return false;
  }
  log('forwarding enabled');
  return true;
}

