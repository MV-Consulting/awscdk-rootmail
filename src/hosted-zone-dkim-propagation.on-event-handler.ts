// eslint-disable-next-line import/no-unresolved
import * as AWSCDKAsyncCustomResource from 'aws-cdk-lib/custom-resources/lib/provider-framework/types';
export const PROP_DOMAIN = 'Domain';

export async function handler(event: AWSCDKAsyncCustomResource.OnEventRequest): Promise<AWSCDKAsyncCustomResource.OnEventResponse> {
  switch (event.RequestType) {
    case 'Create':
    case 'Update':
    case 'Delete':
      console.log(`${event.RequestType} DKIM propagation, doing nothing`);
      return {
        PhysicalResourceId: event.PhysicalResourceId,
      };
  }
}

