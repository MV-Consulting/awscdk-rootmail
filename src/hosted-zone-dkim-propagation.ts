import {
  CustomResource,
  Duration,
  Stack,
  aws_iam as iam,
  aws_lambda as lambda,
} from 'aws-cdk-lib';
import { NodejsFunction } from 'aws-cdk-lib/aws-lambda-nodejs';
import * as cr from 'aws-cdk-lib/custom-resources';
import { NagSuppressions } from 'cdk-nag';
import { Construct, Node } from 'constructs';
import { PROP_DOMAIN } from './hosted-zone-dkim-propagation.on-event-handler';

export interface HostedZoneDKIMPropagationProps {
  readonly domain: string;
  readonly totalTimeToWireDNS?: Duration;
}

export class HostedZoneDKIMPropagation extends Construct {
  constructor(scope: Construct, id: string, props: HostedZoneDKIMPropagationProps) {
    super(scope, id);

    // TODO can we pass in custom cloudwatch log groups? so they are deleted on destroy

    new CustomResource(this, 'Resource', {
      serviceToken: HostedZoneDKIMPropagationProvider.getOrCreate(this, { totalTimeToWireDNS: props.totalTimeToWireDNS }),
      resourceType: 'Custom::HostedZoneDKIMPropagation',
      properties: {
        [PROP_DOMAIN]: props.domain,
      },
    });
  }
}

interface HostedZoneDKIMPropagationProviderProps {
  readonly totalTimeToWireDNS?: Duration;
}

class HostedZoneDKIMPropagationProvider extends Construct {

  /**
   * Returns the singleton provider.
   */
  public static getOrCreate(scope: Construct, props: HostedZoneDKIMPropagationProviderProps) {
    const stack = Stack.of(scope);
    const id = 'rootmail.hosted-zone-dkim-propagation-provider';
    const x = Node.of(stack).tryFindChild(id) as HostedZoneDKIMPropagationProvider
      || new HostedZoneDKIMPropagationProvider(stack, id, props);
    return x.provider.serviceToken;
  }

  private readonly provider: cr.Provider;

  constructor(scope: Construct, id: string, props: HostedZoneDKIMPropagationProviderProps) {
    super(scope, id);

    const isCompleteHandlerFunc = new NodejsFunction(this, 'is-complete-handler', {
      runtime: lambda.Runtime.NODEJS_18_X,
      logRetention: 3,
      timeout: Duration.seconds(30),
      bundling: {
        esbuildArgs: {
          '--packages': 'bundle',
        },
      },
    });

    isCompleteHandlerFunc.addToRolePolicy(
      new iam.PolicyStatement({
        actions: [
          'ses:GetIdentityVerificationAttributes',
          'ses:GetAccountSendingEnabled',
          'ses:GetIdentityDkimAttributes',
          'ses:GetIdentityNotificationAttributes',
        ],
        effect: iam.Effect.ALLOW,
        resources: ['*'],
      }),
    );
    NagSuppressions.addResourceSuppressions(isCompleteHandlerFunc, [
      { id: 'AwsSolutions-IAM4', reason: 'no service role restriction needed' },
      { id: 'AwsSolutions-IAM5', reason: 'wildcards are ok for the ses mail verification' },
    ], true);

    const onEventHandlerFunc = new NodejsFunction(this, 'on-event-handler', {
      runtime: lambda.Runtime.NODEJS_18_X,
      logRetention: 3,
      timeout: Duration.seconds(10),
      bundling: {
        esbuildArgs: {
          '--packages': 'bundle',
        },
      },
    });
    NagSuppressions.addResourceSuppressions(onEventHandlerFunc, [
      { id: 'AwsSolutions-IAM4', reason: 'no service role restriction needed' },
      { id: 'AwsSolutions-IAM5', reason: 'wildcards are ok as the function by itself does nothing' },
    ], true);

    this.provider = new cr.Provider(this, 'hosted-zone-dkim-propagation-provider', {
      isCompleteHandler: isCompleteHandlerFunc,
      queryInterval: Duration.seconds(10),
      totalTimeout: props.totalTimeToWireDNS,
      onEventHandler: onEventHandlerFunc,
    });
    NagSuppressions.addResourceSuppressions(this.provider, [
      { id: 'AwsSolutions-IAM4', reason: 'no service role restriction needed' },
      { id: 'AwsSolutions-IAM5', reason: 'wildcards are ok for the provider as the function has restrictions' },
    ], true);
  };
}

