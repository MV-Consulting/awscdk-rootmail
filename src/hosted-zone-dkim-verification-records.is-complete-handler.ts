// eslint-disable-next-line import/no-unresolved
import * as AWSCDKAsyncCustomResource from 'aws-cdk-lib/custom-resources/lib/provider-framework/types';
import * as AWS from 'aws-sdk';
export const PROP_DOMAIN = 'Domain';
export const ATTR_VERIFICATION_TOKEN = 'VerificationToken';
export const ATTR_DKIM_TOKENS = 'DkimTokens';

// this is fixed to eu-west-1 until SES supports receive more globally (see #23)
const SES = new AWS.SES({ region: 'eu-west-1' });

export interface IsCompleteHandlerResponse {
  IsComplete: boolean;
}

export async function handler(event: AWSCDKAsyncCustomResource.OnEventRequest): Promise<IsCompleteHandlerResponse> {
  const domain = event.ResourceProperties[PROP_DOMAIN];
  let isReady = true;

  console.log(domain);
  console.log(SES);

  return {
    IsComplete: isReady,
  };
}

