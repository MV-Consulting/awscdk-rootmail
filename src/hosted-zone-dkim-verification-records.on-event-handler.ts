// eslint-disable-next-line import/no-unresolved
import * as AWSCDKAsyncCustomResource from 'aws-cdk-lib/custom-resources/lib/provider-framework/types';
import * as AWS from 'aws-sdk';
export const PROP_DOMAIN = 'Domain';
export const ATTR_VERIFICATION_TOKEN = 'VerificationToken';
export const ATTR_DKIM_TOKENS = 'DkimTokens';

const SES = new AWS.SES();

export async function handler(event: AWSCDKAsyncCustomResource.OnEventRequest): Promise<AWSCDKAsyncCustomResource.OnEventResponse> {
  const domain = event.ResourceProperties[PROP_DOMAIN];
  switch (event.RequestType) {
    case 'Create':
    case 'Update': // TODO check for double creation
      let physicalResourceId = event.PhysicalResourceId;
      if (event.RequestType === 'Create') {
        physicalResourceId = event.RequestId;
      }

      console.log(`${event.RequestType}: Do Domain verification and DKIM records for ${event.LogicalResourceId} and domain '${domain}' with PhysicalResourceId '${physicalResourceId}'`);
      const verifyDomainResponse = await SES.verifyDomainIdentity({ Domain: domain }).promise();
      const verificationToken = verifyDomainResponse.VerificationToken;
      console.log(`${event.RequestType}: Got verification token '${verificationToken}' for domain '${domain}'`);

      const verifyDomainDkimResponse = await SES.verifyDomainDkim({ Domain: domain }).promise();
      const dkimTokens = verifyDomainDkimResponse.DkimTokens;
      console.log(`${event.RequestType}: Got DKIM tokens '${dkimTokens}' for domain '${domain}'`);


      return {
        PhysicalResourceId: physicalResourceId,
        Data: {
          [ATTR_VERIFICATION_TOKEN]: verificationToken,
          [ATTR_DKIM_TOKENS]: dkimTokens,
        },
      };
    case 'Delete':
      console.log(`Deleting Domain identity for domain '${domain}' with PhysicalResourceId '${event.PhysicalResourceId}'`);
      const deleteResponse = await SES.deleteIdentity({ Identity: domain }).promise();
      console.log(`Deleted Domain identity for domain '${domain}'`, deleteResponse);
      return {
        PhysicalResourceId: event.PhysicalResourceId,
      };
  }
}

