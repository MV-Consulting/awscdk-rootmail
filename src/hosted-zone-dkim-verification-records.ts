import {
  CustomResource,
  Duration,
  Stack,
  aws_iam as iam,
} from 'aws-cdk-lib';
import * as lambda from 'aws-cdk-lib/aws-lambda-nodejs';
import * as cr from 'aws-cdk-lib/custom-resources';
import { Construct, Node } from 'constructs';
import { ATTR_VERIFICATION_TOKEN, ATTR_DKIM_TOKENS, PROP_DOMAIN } from './hosted-zone-dkim-verification-records.on-event-handler';

export interface HostedZoneDKIMAndVerificationRecordsProps {
  readonly domain: string;
}

export class HostedZoneDKIMAndVerificationRecords extends Construct {
  public readonly verificationToken: string;
  public readonly dkimTokens: string[];

  constructor(scope: Construct, id: string, props: HostedZoneDKIMAndVerificationRecordsProps) {
    super(scope, id);

    const resource = new CustomResource(this, 'Resource', {
      serviceToken: HostedZoneDKIMAndVerificationRecordsProvider.getOrCreate(this),
      resourceType: 'Custom::HostedZoneDKIMAndVerificationRecords',
      properties: {
        [PROP_DOMAIN]: props.domain,
      },
    });

    this.verificationToken = resource.getAttString(ATTR_VERIFICATION_TOKEN);
    this.dkimTokens = resource.getAtt(ATTR_DKIM_TOKENS).toStringList();
  }
}

class HostedZoneDKIMAndVerificationRecordsProvider extends Construct {

  /**
   * Returns the singleton provider.
   */
  public static getOrCreate(scope: Construct) {
    const stack = Stack.of(scope);
    const id = 'superwerker.hosted-zone-dkim-verification-records-provider';
    const x = Node.of(stack).tryFindChild(id) as HostedZoneDKIMAndVerificationRecordsProvider
      || new HostedZoneDKIMAndVerificationRecordsProvider(stack, id);
    return x.provider.serviceToken;
  }

  private readonly provider: cr.Provider;

  constructor(scope: Construct, id: string) {
    super(scope, id);

    this.provider = new cr.Provider(this, 'hosted-zone-dkim-verification-records-provider', {
      onEventHandler: new lambda.NodejsFunction(this, 'on-event-handler', {
        timeout: Duration.seconds(200),
        initialPolicy: [
          new iam.PolicyStatement({
            resources: ['*'],
            actions: [
              'ses:VerifyDomainDkim',
              'ses:VerifyDomainIdentity',
            ],
          }),
        ],
      }),
    });
  };
}