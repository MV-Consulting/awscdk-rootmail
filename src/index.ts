export * from './ses-receive';
export * from './rootmail';