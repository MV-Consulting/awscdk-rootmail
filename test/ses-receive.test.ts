import {
  App,
  Stack,
  aws_s3 as s3,
} from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import { SESReceive, SESReceiveProps } from '../src';

test('ses-receive-stack-create', () => {
  const app = new App();
  const stack = new Stack(app, 'testStack', {
    env: {
      region: 'eu-west-1',
      account: '1234',
    },
  });

  const emailBucket = new s3.Bucket(stack, 'EmailBucket', {
    bucketName: 'test-email-bucket',
    blockPublicAccess: s3.BlockPublicAccess.BLOCK_ALL,
  });

  const testProps: SESReceiveProps = {
    domain: 'example.com',
    subdomain: 'aws',
    emailbucket: emailBucket,
  };

  new SESReceive(stack, 'testSesReceive', testProps);

  const template = Template.fromStack(stack);
  expect(template.toJSON()).toMatchSnapshot();
});