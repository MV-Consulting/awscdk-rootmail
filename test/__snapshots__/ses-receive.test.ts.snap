// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ses-receive-stack-create 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "EmailBucket843A740F": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "BucketName": "test-email-bucket",
        "PublicAccessBlockConfiguration": {
          "BlockPublicAcls": true,
          "BlockPublicPolicy": true,
          "IgnorePublicAcls": true,
          "RestrictPublicBuckets": true,
        },
      },
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Retain",
    },
    "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aFD4BFC8A": {
      "DependsOn": [
        "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRoleDefaultPolicyADDA7DEB",
        "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": "cdk-hnb659fds-assets-1234-eu-west-1",
          "S3Key": "5cc92ed0cea39e2b8de2dbc527dfb5980a3af9564bd1084d840b9787c7d0467e.zip",
        },
        "Handler": "index.handler",
        "Role": {
          "Fn::GetAtt": [
            "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB",
            "Arn",
          ],
        },
        "Runtime": "nodejs18.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRoleDefaultPolicyADDA7DEB": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "logs:PutRetentionPolicy",
                "logs:DeleteRetentionPolicy",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRoleDefaultPolicyADDA7DEB",
        "Roles": [
          {
            "Ref": "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "superwerkersesreceiptrulesetactivationproviderSesReceiptRuleSetActivationCustomResourceRole3C8743C6": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
        "Policies": [
          {
            "PolicyDocument": {
              "Statement": [
                {
                  "Action": [
                    "ses:CreateReceiptRuleSet",
                    "ses:CreateReceiptRule",
                    "ses:SetActiveReceiptRuleSet",
                    "ses:DeleteReceiptRule",
                    "ses:DeleteReceiptRuleSet",
                  ],
                  "Effect": "Allow",
                  "Resource": "*",
                },
              ],
              "Version": "2012-10-17",
            },
            "PolicyName": "AllowSesAccess",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "superwerkersesreceiptrulesetactivationproviderframeworkonEventAEEF5528": {
      "DependsOn": [
        "superwerkersesreceiptrulesetactivationproviderframeworkonEventServiceRoleDefaultPolicy2BA0AECF",
        "superwerkersesreceiptrulesetactivationproviderframeworkonEventServiceRoleA4FA4B79",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": "cdk-hnb659fds-assets-1234-eu-west-1",
          "S3Key": "f2d30cfc360482320a52a4fcde8a70f3569df79ab30be24650fda58eb60052cf.zip",
        },
        "Description": "AWS CDK resource provider framework - onEvent (testStack/superwerker.ses-receipt-ruleset-activation-provider/ses-receipt-ruleset-activation-provider)",
        "Environment": {
          "Variables": {
            "USER_ON_EVENT_FUNCTION_ARN": {
              "Fn::GetAtt": [
                "superwerkersesreceiptrulesetactivationprovideroneventhandler4B6BF571",
                "Arn",
              ],
            },
          },
        },
        "Handler": "framework.onEvent",
        "Role": {
          "Fn::GetAtt": [
            "superwerkersesreceiptrulesetactivationproviderframeworkonEventServiceRoleA4FA4B79",
            "Arn",
          ],
        },
        "Runtime": "nodejs18.x",
        "Timeout": 900,
      },
      "Type": "AWS::Lambda::Function",
    },
    "superwerkersesreceiptrulesetactivationproviderframeworkonEventServiceRoleA4FA4B79": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "superwerkersesreceiptrulesetactivationproviderframeworkonEventServiceRoleDefaultPolicy2BA0AECF": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "lambda:InvokeFunction",
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "superwerkersesreceiptrulesetactivationprovideroneventhandler4B6BF571",
                    "Arn",
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      {
                        "Fn::GetAtt": [
                          "superwerkersesreceiptrulesetactivationprovideroneventhandler4B6BF571",
                          "Arn",
                        ],
                      },
                      ":*",
                    ],
                  ],
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "superwerkersesreceiptrulesetactivationproviderframeworkonEventServiceRoleDefaultPolicy2BA0AECF",
        "Roles": [
          {
            "Ref": "superwerkersesreceiptrulesetactivationproviderframeworkonEventServiceRoleA4FA4B79",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "superwerkersesreceiptrulesetactivationprovideroneventhandler4B6BF571": {
      "DependsOn": [
        "superwerkersesreceiptrulesetactivationproviderSesReceiptRuleSetActivationCustomResourceRole3C8743C6",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": "cdk-hnb659fds-assets-1234-eu-west-1",
          "S3Key": "cf2176897764b1fc368bd19b396ab7672e64078c6b5c363db8dd9bf980650037.zip",
        },
        "Environment": {
          "Variables": {
            "AWS_NODEJS_CONNECTION_REUSE_ENABLED": "1",
          },
        },
        "Handler": "index.handler",
        "Role": {
          "Fn::GetAtt": [
            "superwerkersesreceiptrulesetactivationproviderSesReceiptRuleSetActivationCustomResourceRole3C8743C6",
            "Arn",
          ],
        },
        "Runtime": "nodejs18.x",
        "Timeout": 150,
      },
      "Type": "AWS::Lambda::Function",
    },
    "superwerkersesreceiptrulesetactivationprovideroneventhandlerLogRetention4030C654": {
      "Properties": {
        "LogGroupName": {
          "Fn::Join": [
            "",
            [
              "/aws/lambda/",
              {
                "Ref": "superwerkersesreceiptrulesetactivationprovideroneventhandler4B6BF571",
              },
            ],
          ],
        },
        "RetentionInDays": 3,
        "ServiceToken": {
          "Fn::GetAtt": [
            "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aFD4BFC8A",
            "Arn",
          ],
        },
      },
      "Type": "Custom::LogRetention",
    },
    "testSesReceiveOpsSantaFunctionRoleE4ADD3F6": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
        "Policies": [
          {
            "PolicyDocument": {
              "Statement": [
                {
                  "Action": "s3:GetObject",
                  "Effect": "Allow",
                  "Resource": {
                    "Fn::Join": [
                      "",
                      [
                        {
                          "Fn::GetAtt": [
                            "EmailBucket843A740F",
                            "Arn",
                          ],
                        },
                        "/RootMail/*",
                      ],
                    ],
                  },
                },
                {
                  "Action": "ssm:CreateOpsItem",
                  "Effect": "Allow",
                  "Resource": "*",
                },
                {
                  "Action": "ssm:PutParameter",
                  "Effect": "Allow",
                  "Resource": {
                    "Fn::Join": [
                      "",
                      [
                        "arn:",
                        {
                          "Ref": "AWS::Partition",
                        },
                        ":ssm:eu-west-1:1234:parameter/superwerker/*",
                      ],
                    ],
                  },
                },
              ],
              "Version": "2012-10-17",
            },
            "PolicyName": "OpsSantaFunctionRolePolicy",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "testSesReceiveSESReceiptRuleSetActivation4F766837": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "Domain": "example.com",
        "EmailBucketName": {
          "Ref": "EmailBucket843A740F",
        },
        "OpsSantaFunctionArn": {
          "Fn::GetAtt": [
            "testSesReceiveopssantahandlerEF47BA3E",
            "Arn",
          ],
        },
        "RulesetSettleTimeSeconds": 120,
        "ServiceToken": {
          "Fn::GetAtt": [
            "superwerkersesreceiptrulesetactivationproviderframeworkonEventAEEF5528",
            "Arn",
          ],
        },
        "Subdomain": "aws",
      },
      "Type": "Custom::SESReceiptRuleSetActivation",
      "UpdateReplacePolicy": "Delete",
    },
    "testSesReceiveopssantahandlerEF47BA3E": {
      "DependsOn": [
        "testSesReceiveOpsSantaFunctionRoleE4ADD3F6",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": "cdk-hnb659fds-assets-1234-eu-west-1",
          "S3Key": "26caa3dcf91fb0925a0d4b2da8aec28d4ace1938bc847ad47802fb3782a8c883.zip",
        },
        "Environment": {
          "Variables": {
            "AWS_NODEJS_CONNECTION_REUSE_ENABLED": "1",
            "EMAIL_BUCKET": {
              "Ref": "EmailBucket843A740F",
            },
            "EMAIL_BUCKET_ARN": {
              "Fn::GetAtt": [
                "EmailBucket843A740F",
                "Arn",
              ],
            },
            "ROOTMAIL_DEPLOY_REGION": "eu-west-1",
          },
        },
        "Handler": "index.handler",
        "Role": {
          "Fn::GetAtt": [
            "testSesReceiveOpsSantaFunctionRoleE4ADD3F6",
            "Arn",
          ],
        },
        "Runtime": "nodejs18.x",
        "Timeout": 60,
      },
      "Type": "AWS::Lambda::Function",
    },
    "testSesReceiveopssantahandlerLogRetentionFF512AB6": {
      "Properties": {
        "LogGroupName": {
          "Fn::Join": [
            "",
            [
              "/aws/lambda/",
              {
                "Ref": "testSesReceiveopssantahandlerEF47BA3E",
              },
            ],
          ],
        },
        "RetentionInDays": 3,
        "ServiceToken": {
          "Fn::GetAtt": [
            "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aFD4BFC8A",
            "Arn",
          ],
        },
      },
      "Type": "Custom::LogRetention",
    },
    "testSesReceiveopssantahandlerOpsSantaFunctionSESPermissionsBC82EBA8": {
      "Properties": {
        "Action": "lambda:InvokeFunction",
        "FunctionName": {
          "Fn::GetAtt": [
            "testSesReceiveopssantahandlerEF47BA3E",
            "Arn",
          ],
        },
        "Principal": "ses.amazonaws.com",
        "SourceAccount": "1234",
      },
      "Type": "AWS::Lambda::Permission",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
